### PR TITLE
Implement image checksum verification for svirt

### PR DESCRIPTION
**Pull request by mnowaksuse**
https://trello.com/c/ATxUsjcv/252-p3-enhance-image-download-mechanism-on-hyper-v

Check image SHA-256 checksum on svirt.

Validation runs:
* jeos-extratest@svirt-hyperv: http://nilgiri.suse.cz/tests/1659#step/data_integrity/3
* jeos-main@svirt-xen-hvm: http://nilgiri.suse.cz/tests/1654#step/data_integrity/3
* allpatterns@svirt-xen-pv: http://nilgiri.suse.cz/tests/1656#step/data_integrity/3
* lvm+RAID1@svirt-hyperv: http://nilgiri.suse.cz/tests/1660#step/data_integrity/3
* default_install@svirt-hyperv2012r2: http://nilgiri.suse.cz/tests/1664#step/data_integrity/3
* default@64bit: http://nilgiri.suse.cz/tests/1657#step/data_integrity/1
* jeos-main-de_DE@64bit-virtio-vga: http://nilgiri.suse.cz/tests/1662#step/data_integrity/1

### CODE REVIEW

File: lib/data_integrity_utils.pm [hunk 1]
```diff
@@ -32,9 +32,21 @@ sub verify_checksum {
         my $checksum = get_required_var $image;
         $image =~ s/CHECKSUM_//;
         my $image_path = get_required_var $image;
+        my $dir_path;
```
Comment (jknphy) [inline lib/data_integrity_utils.pm]: $dir_path is re-declared here. If you don't want to pass the path from the caller I would suggest to create function get_path and delete the parameter.

File: lib/data_integrity_utils.pm [hunk 2]
```diff
@@ -32,9 +32,21 @@ sub verify_checksum {
...
+        $dir_path = '/var/lib/libvirt/images/' if check_var('BACKEND', 'svirt');
+        $dir_path = get_var('HYPERV_DISK', 'D:') . '\\cache\\' if check_var('VIRSH_VMM_FAMILY', 'hyperv');
         $image_path = $dir_path . basename($image_path) if $dir_path;
-        my $digest = digest_file_hex($image_path, "SHA-256");
-        record_info("$image Ok", "$image_path: Ok") && next if $checksum eq $digest;
+        my $digest;
```
Comment (jknphy) [inline lib/data_integrity_utils.pm]: Same thing for this block, create a function that return the digest and encapsulate from line 39-48 or (I like more this option): `my $digest = check_var('BACKEND', 'svirt')) ? get_digest_svirt : digest_file_hex($image_path, "SHA-256");

Comment (jknphy) [reply lib/data_integrity_utils.pm]: you could not find `digest_file_hex` in the worker for svirt? We added here: https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/cpanfile#L16

**Comment by jknphy** [comment]
Nevertheless, before going forward with the changes due to the [test is not scheduled](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/5982/files) anymore  and the original ticket is [blocked](https://progress.opensuse.org/issues/41414), let me check with them again the status:

File: lib/data_integrity_utils.pm (hunk 2 above)
Comment (mnowaksuse) [reply lib/data_integrity_utils.pm]: For svirt `digest_file_hex` checks the image on worker, but the image needs to be checked on the host as on svirt worker is not a host.

**Comment by mnowaksuse** [comment]
Thanks. Re-validated and added two new runs on qemu backend.

File: lib/data_integrity_utils.pm [hunk 3]
```diff
@@ -19,22 +19,43 @@ use testapi;
...
+=cut
+sub get_image_digest {
+    my ($image_path) = shift;
+
+    my $digest;
+    if (check_var('BACKEND', 'svirt')) {
+        $digest = console('svirt')->get_cmd_output("sha256sum $image_path");
+        # On Hyper-V the hash starts with '\'
+        my $start = check_var('VIRSH_VMM_FAMILY', 'hyperv') ? 1 : 0;
+        $digest = substr $digest, $start, 64;    # extract SHA256 from the output
```
Comment (jknphy) [inline lib/data_integrity_utils.pm]: you can return the $digest here (as it is better in Perl not declare variables if it is not really needed, for maintenance)

File: lib/data_integrity_utils.pm [hunk 4]
```diff
@@ -19,22 +19,43 @@ use testapi;
...
+    }
+    else {
+        $digest = digest_file_hex($image_path, "SHA-256");
```
Comment (jknphy) [inline lib/data_integrity_utils.pm]: return here too

File: lib/data_integrity_utils.pm [hunk 5]
```diff
@@ -19,22 +19,43 @@ use testapi;
...
+    }
+    return $digest;
```
Comment (jknphy) [inline lib/data_integrity_utils.pm]: ... and delete this line.

File: lib/data_integrity_utils.pm [hunk 6]
```diff
@@ -19,22 +19,43 @@ use testapi;
...
 sub verify_checksum {
-    my ($dir_path) = shift;    # for backends other than qemu, image directory path needs to be set
+    my ($dir_path) = shift;
 
     diag "Comparing data integrity calculated with SHA256 digest against checksum from IBS/OBS via rsync.pl";
     foreach my $image (grep { /^CHECKSUM_/ } keys %bmwqemu::vars) {
         my $checksum = get_required_var $image;
         $image =~ s/CHECKSUM_//;
         my $image_path = get_required_var $image;
+        $dir_path = '/var/lib/libvirt/images/' if check_var('BACKEND', 'svirt');
```
Comment (jknphy) [inline lib/data_integrity_utils.pm]: I would suggest to do this two assignation from outside as the method is prepared for accept $dir_path values. Or if you don't want to use it, remove the parameter and extract similar method that you did for `get_image_digest`

File: lib/data_integrity_utils.pm (hunk 3 above)
Comment (mnowaksuse) [reply lib/data_integrity_utils.pm]: I find it a matter of taste. I'll keep it like this. Thanks.

File: lib/data_integrity_utils.pm (hunk 4 above)
Comment (mnowaksuse) [reply lib/data_integrity_utils.pm]: Ditto.

File: lib/data_integrity_utils.pm (hunk 5 above)
Comment (mnowaksuse) [reply lib/data_integrity_utils.pm]: Ditto.

File: lib/data_integrity_utils.pm (hunk 6 above)
Comment (mnowaksuse) [reply lib/data_integrity_utils.pm]: OK, I pushed them to the test itself.

File: tests/installation/data_integrity.pm [hunk 7]
```diff
@@ -16,7 +16,10 @@ use testapi;
 use data_integrity_utils 'verify_checksum';
 
 sub run {
-    verify_checksum;
+    my $dir_path;
```
Comment (jknphy) [inline tests/installation/data_integrity.pm]: Without an orphan declaration (but it is again a matter of taste :) ). Feel free to leave it as it is.
```
$dir_path = check_var('BACKEND', 'svirt') ? '/var/lib/libvirt/images/'
           : check_var('VIRSH_VMM_FAMILY', 'hyperv') ? get_var('HYPERV_DISK', 'D:') . '\\cache\\' : undef;
```

Comment (foursixnine) [reply tests/installation/data_integrity.pm]: A ternary operator will make the code more difficult to read, just a comment would do.

**Comment by okurz** [comment]
@mnowaksuse I failed to actually find the module "data_integrity" scheduled in production. https://openqa.suse.de/tests/2256068 for example does not have it.

**Comment by mnowaksuse** [comment]
@okurz Indeed, it's not scheduled for anyone. This PR is to make sure, that once enabled for qemu backend, it can be enabled for svirt as well - the only difference is likely where the test is scheduled. For qemu you may schedule it around `isosize` test module, for svirt it needs to be started after `bootloader_svirt`.

**Pull request merged**